### PR TITLE
Alignment to pulse constraints

### PR DESCRIPTION
**Pull request by nkanazawa1989**
### Summary

This is a follow-up PR of #6673 due to the specification format update of the backend configuration. 

### Current format
```
backend.configuration().alignment
```
This returns a single integer value.

### New format
```
backend.configuration().pulse_constraints
```
This returns a dictionary of multiple integer values. Namely, `granularity`, `min_length`, `control_alignment`, `measure_alignment`. These names are TBD. These properties are generally necessary information to validate pulse gate experiments on front end.

### Details and comments

This PR will be updated once the spec is fixed.

### CODE REVIEW

File: qiskit/compiler/transpiler.py [hunk 1]
```diff
@@ -148,21 +149,25 @@ def transpile(
...
-            If the backend doesn't have any restriction on the alignment,
-            then ``alignment`` is None and no adjustment will be performed.
-
-            .. note::
-
-                Note that currently this instruction time optimization is performed only for
-                ``measure`` instructions. Here we assume a hardware whose
-                measurements should start at a time which is a multiple of the alignment value.
-
+        pulse_constraints: An optional control hardware restrictions instruction time resolution.
```
Comment (ajavadia) [inline qiskit/compiler/transpiler.py:152]:
```suggestion
        pulse_constraints: An optional control hardware restriction on instruction time resolution.
```

File: qiskit/compiler/transpiler.py [hunk 2]
```diff
@@ -856,14 +861,27 @@ def _parse_output_name(output_name, circuits):
...
-    if not isinstance(alignment, int) or alignment == 0:
-        raise TranspilerError(f"Alignment should be nonzero integer value. Not {alignment}.")
+    # validation
+    fields = ["granularity", "min_length", "control_alignment", "measure_alignment"]
+    for field in fields:
+        constraint_value = getattr(pulse_constraints, field, None)
+        if not isinstance(constraint_value, int) or constraint_value < 1:
+            raise TranspilerError(
+                f"Pulse constraint {field} should be nonzero integer. Not {constraint_value}."
+            )
```
Comment (ajavadia) [inline qiskit/compiler/transpiler.py:885]: I think this validation should go inside PulseConstraints itself during construction. Also make it fast when there's no constraint.

**Review by ajavadia — COMMENTED** [review]
overall looks good. Will there be a follow-on transpiler pass for `AlignGates`?

File: qiskit/compiler/transpiler.py (hunk 2 above)
Comment (nkanazawa1989) [reply qiskit/compiler/transpiler.py]: Thanks, this is moved to the constructor https://github.com/Qiskit/qiskit-terra/pull/6705/commits/0c38fcd0dac96e900713b5301cf8b791f495f86f

File: qiskit/transpiler/passmanager_config.py [hunk 3]
```diff
@@ -55,7 +55,7 @@ def __init__(
                 (1.0=no approximation, 0.0=maximal approximation)
             seed_transpiler (int): Sets random seed for the stochastic parts of
                 the transpiler.
-            alignment (int): Hardware instruction alignment restriction.
+            timing_constraints (PulseConstraints): Hardware time alignment restrictions.
```
Comment (ajavadia) [inline qiskit/transpiler/passmanager_config.py:58]:
```suggestion
            timing_constraints (TimingConstraints): Hardware time alignment restrictions.
```

**Review by ajavadia — COMMENTED** [review]
LGTM

**Review by ajavadia — APPROVED** [review]

**Pull request merged**